File: ml-benchmark/synduce/constraints/sortedlist/sndmax_empty_base_case.f
Inductive CList = empty Unit | elt Int | concat {CList, CList};
Inductive List = nil Unit | cons {Int, List};

cat = fix (
  \f: List -> List -> List. \a: List. \b: List.
  match a with
    nil _ -> b
  | cons {h, t} -> cons {h, (f t b)}
  end
);

repr = fix (
  \f: CList -> List. \xs: CList.
  match xs with
    empty _ -> nil unit
  | elt w -> cons {w, nil unit}
  | concat {l, r} -> cat (f l) (f r)
  end
);

all_pos = fix (
  \f: CList -> Bool. \c: CList.
  match c with
    empty _ -> true
  | elt w -> > w 0
  | concat {l, r} -> and (f l) (f r)
  end
);

geq_head = \x: Int. \xs: List.
  match xs with
    nil _ -> true
  | cons {h, t} -> >= x h
  end;

is_sorted = fix (
  \f: List -> Bool. \xs: List.
  match xs with
    nil _ -> true
  | cons {h, t} -> and (geq_head h t) (f t)
  end
);

min = \x: Int. \y: Int. if < x y then x else y;
max = \x: Int. \y: Int. if > x y then x else y;

spec = \xs: List. (fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    nil _ -> {0, 0}
  | cons {h, t} ->
    let res = f t in
      {max res.1 h, max res.2 (min res.1 h)}
  end
) xs).2;

target = fix (
  \f: CList -> Compress CList. \c: CList.
  match c with
    empty _ -> c
  | elt w -> c
  | concat {l, r} -> concat {f l, f r}
  end
);

main = \c: CList.
  if is_sorted (repr c) then spec (repr (target c)) else 0;

File: ml-benchmark/synduce/constraints/sortedlist/sndmax_empty_base_case.ml
type clist = Empty | Elt of int | Concat of clist * clist
type ilist = Nil | Cons of int * ilist

let rec cat a b =
  match a with
  | Nil -> b
  | Cons (h, t) -> Cons (h, cat t b)

let rec repr xs =
  match xs with
  | Empty -> Nil
  | Elt w -> Cons (w, Nil)
  | Concat (l, r) -> cat (repr l) (repr r)

let rec all_pos c =
  match c with
  | Empty -> true
  | Elt w -> w > 0
  | Concat (l, r) -> all_pos l && all_pos r

let geq_head x xs =
  match xs with
  | Nil -> true
  | Cons (h, t) -> x >= h

let rec is_sorted xs =
  match xs with
  | Nil -> true
  | Cons (h, t) -> geq_head h t && is_sorted t

let min x y = if x < y then x else y
let max x y = if x > y then x else y

let spec xs =
  let rec f xs =
    match xs with
    | Nil -> (0, 0)
    | Cons (h, t) ->
      let res = f t in
      match res with
      | (r1, r2) -> (max r1 h, max r2 (min r1 h))
  in
  let res = f xs in
  match res with
  | (_, s) -> s

val target: clist -> clist compress
let rec target c =
  match c with
  | Empty -> c
  | Elt w -> c
  | Concat (l, r) -> Concat (target l, target r)

let program c =
  if is_sorted (repr c) then spec (repr (target c)) else 0